Tutorials/WinForms/Classic/bind-the-data-grid-to-large-data-source.md

```diff
@@ -0,0 +1,301 @@
+    * ProductName, TextEdit
+    * OrderDate, DateEdit
+    * Freight, CalcEdit
+    * Customer, LookupEdit 
+* Click the **Finish** button, delete the **First Name** and **Last Name** layout items, reorder the rest layout items to create a layout similar to the `EditCustomerForm`, and save the changes.
+* Add the `XPBindingSource` component to the `Form` and rename it to **CustomersBindingSource**.
+* Rebuild the project.
+* Set the `CustomersBindingSource.ObjectClassInfo` property to **DxSample.DataAccess.Customer**.
+* Set the `CustomersBindingSource.DisplayableProperties` property to **Oid;ContactName**.
+* Select the `lookUpEdit1` editor placed near the **Customer** label and click the [smart-tag](https://docs.microsoft.com/en-us/dotnet/framework/winforms/controls/performing-common-tasks-using-smart-tags-on-wf-controls) glyph. 
+* Click the **Data Source** item to open the drop-down list, find the `CustomersBindingSource` component under the **Other Data Sources > EditOrderForm List Instances** node, and select it. 
+* Select the **ContactName** and **Oid** properties as the **Display Member** and **Value Member** if they were not selected automatically.
+* Open the code editor and change the code as follows:
+    <details>
+        <summary>Click to expand</summary>   
+
+    ```csharp
+    using DevExpress.Xpo;
+    using DevExpress.Xpo.DB.Exceptions;
+    using DevExpress.XtraEditors;
+    using DxSample.DataAccess;
+    using System;
+    using System.Windows.Forms;
+
+    namespace DxSample.Forms {
+        public partial class EditOrderForm : XtraForm {
+            public EditOrderForm() {
+                InitializeComponent();
+            }
+            public EditOrderForm(int? orderId) : this() {
+                OrderID = orderId;
+            }
+            public int? OrderID { get; private set; }
+            protected UnitOfWork UnitOfWork { get; private set; }
+            private void EditCustomerForm_Load(object sender, EventArgs e) {
+                Reload();
+            }
+
+            private void Reload() {
+                UnitOfWork = new UnitOfWork();
+                if (OrderID.HasValue)
+                    OrderBindingSource.DataSource = UnitOfWork.GetObjectByKey<Order>(OrderID.Value);
+                else OrderBindingSource.DataSource = new Order(UnitOfWork);
+                CustomersBindingSource.DataSource = new XPCollection<Customer>(UnitOfWork);
+            }
+
+            private void btnSave_Click(object sender, EventArgs e) {
+                try {
+                    UnitOfWork.CommitChanges();
+                    OrderID = ((Order)OrderBindingSource.DataSource).Oid;
+                    Close();
+                } catch (LockingException) {
+                    XtraMessageBox.Show(this, "The record was modified or deleted by another user. Please click the Reload button and try again.", "XPO Tutorial", MessageBoxButtons.OK, MessageBoxIcon.Stop);
```

Comment:
```suggestion
                    XtraMessageBox.Show(this, "The record was modified or deleted. Please click the Reload button and try again.", "XPO Tutorial", MessageBoxButtons.OK, MessageBoxIcon.Stop);
```

Reply:
Added this change to the updated version of the example, too.